ADO.NET/OAuthSampleApp/OAuthSampleApp.cs

```diff
@@ -0,0 +1,217 @@
+        cmd.ExecuteNonQuery();
+        cmd.CommandText = "DROP AUTHENTICATION IF EXISTS adooauth CASCADE";
+        cmd.ExecuteNonQuery();
+        cmd.CommandText = "CREATE USER " + USER;
+        cmd.ExecuteNonQuery();
+        cmd.CommandText = "CREATE AUTHENTICATION adooauth METHOD 'oauth'" + GetAuthHost();
+        cmd.ExecuteNonQuery();
+        cmd.CommandText = "GRANT AUTHENTICATION adooauth TO " + USER;
+        cmd.ExecuteNonQuery();
+        cmd.CommandText = "GRANT ALL ON SCHEMA PUBLIC TO " + USER;
+        cmd.ExecuteNonQuery();
+        cmd.CommandText = "ALTER AUTHENTICATION adooauth SET client_id = '" + CLIENT_ID + "'";
+        cmd.ExecuteNonQuery();
+        cmd.CommandText = "ALTER AUTHENTICATION adooauth SET client_secret = '" + CLIENT_SECRET + "'";
+        cmd.ExecuteNonQuery();
+        cmd.CommandText = "ALTER AUTHENTICATION adooauth SET introspect_url = '" + INTROSPECT_URL + "'";
+        cmd.ExecuteNonQuery();
+    }
+
+    // removes the oauth user and closes the admin connection
+    public static async Task TearDown()
+    {
+        string USER = ConfigurationManager.AppSettings["User"];
+        VerticaCommand cmd = new VerticaCommand("DROP USER IF EXISTS " + USER + " CASCADE", vConnection);
+        cmd.ExecuteNonQuery();
+        cmd.CommandText = "DROP AUTHENTICATION IF EXISTS adooauth CASCADE";
+        cmd.ExecuteNonQuery();
+
+        vConnection.Close();
+    }
+
+    // if no access token is set, obtains and sets first access and refresh tokens
+    // uses the password grant flow
+    static async Task EnsureAccessToken()
+    {
+        string accessToken = Environment.GetEnvironmentVariable(OAUTH_ACCESS_TOKEN_VAR_STRING, EnvironmentVariableTarget.User);
+
+        if (string.IsNullOrEmpty(accessToken))
+        {
+            // Obtain first access token
+            Console.WriteLine("Access token not found. Obtaining first access token from IDP.");
+            await DoPasswordGrant();
+        }
+
+        SetAccessToken();
+    }
+
+    // the access token must be set in the connection string
+    static void SetAccessToken()
+    {
+        string accessToken = Environment.GetEnvironmentVariable(OAUTH_ACCESS_TOKEN_VAR_STRING, EnvironmentVariableTarget.User);
+        connectionStringBuilder["OAuthAccessToken"] = accessToken;
+    }
+
+    // OAuth user attempts to connect with the access token
+    // If it fails, we try to get a new access token with the refresh token
+    // If the refresh token is expired, we get new access and refresh tokens
+    public static async Task ConnectToDatabase()
+    {
+        // ADO.NET requires a user in the connection string
+        // It is ignored by the driver, and only the access token is used
+        // The user can be set to any value (except the admin) and it will connect
+        connectionStringBuilder.User = ConfigurationManager.AppSettings["User"];
+        int retryCount = 0;

```

Comment:
This seems a bit misleading. At the start of the first try, the retry count is 1 which makes it seem like we have already retried when really we are just starting the first retry. Maybe just numAttempts or something like that

Reply:
Sounds good